### PR TITLE
Fix for Windows

### PR DESCRIPTION
**Pull request by juntalis**
Fix for [Issue #1](https://github.com/dejw/vip/issues/1)

### CODE REVIEW

**Comment by dejw** [comment]
Actually, it would be nice to have a test case for this issue as well.

Can you implement this as well?

Thanks.

**Comment by juntalis** [comment]
Sorry about that. Got a bit preoccupied and didn't notice your comment. Sure, I'll go ahead and throw something together over the weekend.

**Comment by dejw** [comment]
Make sure that your test runs on Windows only and Travis looks good (https://travis-ci.org/dejw/vip/pull_requests).

Thanks.

**Comment by juntalis** [comment]
While writing the tests, I noticed a couple of other issues:
- Two of the existing tests fail under Windows due to how the path is constructed in the test.
- The requirements functionality actually throws an error on Windows. (Didn't notice it until now, since I've been using vip when creating a new project in empty folders) It occurs when execute_virtualenv_command is unable to find the pip executable.

I've already implemented fixes for both of these (and another os-specific test case for the fix to execute_virtualenv_command), but wasn't sure how you'd prefer them. Some developers prefer separate issues/pull requests on each actual issue, and others prefer having to do less work in the present.

Lastly, if you'd like, I wouldn't mind helping to maintain Windows-compatibility as you go forward with the project. Due to work, I'm on Windows most of the time nowadays, but since I have a few projects targeting both Linux and Windows, I understand it gets kind of laborious after a while. (Especially when you have to open a virtual machine or switch partitions just to do some testing) I can also see why you didn't think it'd be used on Windows, considering that it's not really the normal Windows way of doing things. The way that I see it, though, since I'm already initializing a git repository in the root folder of every new project, it feels a lot more natural to utilize vip over the default virtualenv script.(Plus, since I use a modified version of [PyCmd](https://sourceforge.net/projects/pycmd/) as a replacement for command prompt, I've already set it up to automatically activate the virtual environment whenever it detects the vip folder)

**Comment by dejw** [comment]
I prefer one pull request for all fixes. 

I will be more than happy to allow you to maintain this project on Windows side.

Personally, I mainly use vip to maintain dependencies for my projects, constantly using `--init` flag to install all the libraries, and `vip python filename` to run something. That is pretty much it, so that is why the implementation progress have moved barely nowhere recently.

I don't like the idea that I have to alter my shell environment to use Python inside the virtualenv, that is why I decided not to integrate my tool with the script itself. Or maybe you thought about different kind of integration? How is it used on Windows?

It would be nice to know, what are your use-cases exactly? What is the work-flow with vip?

**Comment by juntalis** [comment]
Well up until this weekend, I was primarily just using vip for initially setting up new projects. There's tons of other systems to facilitate that functionality, but I've always kind of preferred a minimalist approach whenever possible. 

Anyways normally, with the stock virtualenv package, that process might consist of the following steps for me:

```
virtualenv --no-site-packages projectdir
cd projectdir
Scripts\activate.bat
mkdir src
cd src
echo argparse>requirements.txt
echo unittest2>>requirements.txt
pip install -r requirements.txt
git --init
```

I probably very well could've eliminated some of those steps by creating the folder, initializing the repo, and setting up the virtual environment in the same folder as the source, but to be honest, using a "." prefixed hidden folder for the virtual environment never really occurred to me until I started using vip. Anyways, the process now became:

```
mkdir projectdir
cd projectdir
echo argparse>requirements.txt
echo unittest2>>requirements.txt
vip --init && git --init
.vip\Scripts\activate.bat
```

Doesn't seem like all that much of difference, but the process does feel a lot smoother.

As far as the shell integration goes, I wasn't suggesting using PyCmd as your shell or anything like that - I was pointing at the most recent usage case for vip. PyCmd is basically a Windows-specific python project that wraps around the built-in Command Prompt (Window's "shell") and adds additional functionality such as better tab completion, more customization, and a few other things that you might find in a real shell. By adding a hook to PyCmd's internal cd function, I was able to integrate in functionality so that when I enter a folder and vip detects it as a virtual environment, it automatically activates that virtual environment without any input from me. When I leave the folder, it then deactivates the virtual environment, and I go back to using the global python interpreter. This saves me a lot of trouble, as it allows me to jump back and forth between different projects in the same console without any hassle.

Basically, it'd be the equivalent of doing something like this in bash:

```
function cd() {
    builtin cd "$@"
    local VIP_DIR=$("vip -l")
    if [ $? -eq 0 ]; then
        . "$VIP_DIR/bin/activate"
    fi
}
```

My bash scripting is a little rusty though, so I'm not sure how you'd automatically deactivate. (Only way I can think of is to use an exported environment variable, and check for that, but you'd probably know a better way if there was one)

**Comment by dejw** [comment]
Thanks for the comment!

About the auto (de)activation feature: I have seen a project that utilized that thing and it worked pretty well (but I forgot the name). I will try to find it, so maybe I can integrate this feature into vip.

On simplifying things: it would be nice to have a possibility to setup some templates like in your example, say:

```
[template:basic]
argparse = latest
unittest2 = latest
```

and then simply type: `vip --init basic` to install both packages to latest version and initialize `requirements.txt` file.

Also the whole tool would shrink to just one command - `--init` or something similar - which would update the state of the virtualenv, installing, removing updating packages when necessary.

**Comment by juntalis** [comment]
I could definitely see a benefit in that kind of functionality. Adding some small functionality for implementing init hooks might help too, though it might add a bit too much extra baggage if you're looking to keep vip small and lean.

Something like:

```
[template:basic]
argparse = latest
unittest2 = latest
[hooks:basic]
init = 'git --init', 'git ignore *.py[co] build dist'
```

**Comment by dejw** [comment]
This looks like a good idea. Unfortunately, I do not know when I am going to implement anything, since I got stuck with something else. Also, I use vip very rarely recently.

**Comment by dejw** [comment]
I recently fixed some major issues, and bumped version to 0.1.3. Unfortunately, I do not know if it breaks Windows compatibility.

**Comment by juntalis** [comment]
Hey,

Updated my local copy and ran through all the tests, then set up a couple of test environment. Everything looks to be working correctly. Also, I completely understand where you're coming from on your previous post. Unfortunately, it's the same reason so many of my repositories contain personal projects that remain unfinished to this day.

**Comment by dejw** [comment]
Everything works, indeed.

I managed to setup a virtual machine with Windows on it, and fix the code ;-)

Too bad that Travis does not have Windows support yet, it would be useful to be able to run tests in an automatic way instead of manual.

**Comment by juntalis** [comment]
Yeah, definitely. [From the looks of it](https://github.com/travis-ci/travis-ci/issues/216), they want to expand support for testing Windows and Mac OS X, as well. If that pans out, it should definitely make things a bit easier.
